provider/azure/instancetype.go

```diff
@@ -17,88 +20,437 @@ import (
+	// ones, so we don't want to be providing them as
+	// a default. This is achieved by costing them at
+	// a higher price, even though they are cheaper
+	// in reality.
+	"Basic_A0",
+	"Basic_A1",
+	"Basic_A2",
+	"Basic_A3",
+	"Basic_A4",
+}
+
 // newInstanceType creates an InstanceType based on a VirtualMachineSize.
 func newInstanceType(size compute.VirtualMachineSize) instances.InstanceType {
-	// We're not doing real costs for now; just made-up, relative
-	// costs, to ensure we choose the right VMs given matching
-	// constraints. This was based on the pricing for West US,
-	// and assumes that all regions have the same relative costs.
-	//
-	// DS is the same price as D, but is targeted at Premium Storage.
-	// Likewise for GS and G. We put the premium storage variants
-	// directly after their non-premium counterparts.
-	machineSizeCost := []string{
-		"Standard_A0",
-		"Standard_A1",
-		"Standard_D1",
-		"Standard_DS1",
-		"Standard_D1_v2",
-		"Standard_A2",
-		"Standard_D2",
-		"Standard_DS2",
-		"Standard_D2_v2",
-		"Standard_D11",
-		"Standard_DS11",
-		"Standard_D11_v2",
-		"Standard_A3",
-		"Standard_D3",
-		"Standard_DS3",
-		"Standard_D3_v2",
-		"Standard_D12",
-		"Standard_DS12",
-		"Standard_D12_v2",
-		"Standard_A5", // Yes, A5 is cheaper than A4.
-		"Standard_A4",
-		"Standard_A6",
-		"Standard_G1",
-		"Standard_GS1",
-		"Standard_D4",
-		"Standard_DS4",
-		"Standard_D4_v2",
-		"Standard_D13",
-		"Standard_DS13",
-		"Standard_D13_v2",
-		"Standard_A7",
-		"Standard_A10",
-		"Standard_G2",
-		"Standard_GS2",
-		"Standard_D5_v2",
-		"Standard_D14",
-		"Standard_DS14",
-		"Standard_D14_v2",
-		"Standard_A8",
-		"Standard_A11",
-		"Standard_G3",
-		"Standard_GS3",
-		"Standard_A9",
-		"Standard_G4",
-		"Standard_GS4",
-		"Standard_GS5",
-		"Standard_G5",
-
-		// Basic instances are less capable than standard
-		// ones, so we don't want to be providing them as
-		// a default. This is achieved by costing them at
-		// a higher price, even though they are cheaper
-		// in reality.
-		"Basic_A0",
-		"Basic_A1",
-		"Basic_A2",
-		"Basic_A3",
-		"Basic_A4",
-	}
-
 	// Anything not in the list is more expensive that is in the list.
 	cost := len(machineSizeCost)
 	sizeName := to.String(size.Name)
+
+	// Actual instance type names often are suffixed with _v3, _v4 etc. We always
+	// prefer the highest version number.
+	namePart := instSizeVersionRegexp.ReplaceAllString(sizeName, "$name")
+	vers := 0
+	if namePart == "" {
+		namePart = sizeName
+	} else {
+		versStr := instSizeVersionRegexp.ReplaceAllString(sizeName, "$version")
+		if versStr != "" {
+			vers, _ = strconv.Atoi(versStr)
+		}
+	}
 	for i, name := range machineSizeCost {
-		if sizeName == name {
-			cost = i
+		if namePart == name {
+			// Space out the relative costs and make a small subtraction
+			// so the higher versions of the same instance have a lower cost.
+			cost = 100*i - vers
 			break
 		}
 	}
 	if cost == len(machineSizeCost) {
```

Comment:
this is wrong, lets set the cost initially to -1 then check that here, updating cost to len(machineSizeCost)*100

Reply:
done